### ISSUE TITLE
Add ability to build/run maint-1.0 tests on pm-cpu with GNU

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/E3SM#5280]
I have some changes that allow choice tests in maint-1.0 to build/run on pm-cpu. `ndk/maint-1.0/add-pm-cpu`

These run routinely on cori-knl:
```
SMS_Ld2.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.cori-knl_intel.allactive-v1cmip6
SMS_Ld2.ne30_oECv3_ICG.A_WCYCL20TRS_CMIP6.cori-knl_intel.allactive-v1cmip6
SMS_Ln5.ne30_ne30.FC5AV1C-L.cori-knl_intel.cam-cosplite
```

And I can run them on pm-cpu with basic machine file changes and a few other mods.

There are 2 changes in files under submodules that are not easy to change.
These are changes needed with higher version of GNU (all that is avail on PM)

Need a change here to build. this is already done in master, but there are other changes we don't need --
can just update source with the interface changes needed
`components/clm/src/external_models/sbetr/src/betr/betr_math/ODEMod.F90`

needed to work-around issue with solver during runtime. this change modifies compiler
flags for ALL MPAS sources which is not necessary, but unable to only affect one source in MPAS
in current master, we build one MPAS source file with a compiler flag to turn off very specific optimization,
however, in maint-1.0, the CMake logic is different and I can't do the same thing in Depends.
But what I can do is add a flag here (which may need more guards) that turns off the optimization
for all MPAS source files.
`components/mpas-source/src/Makefile.in.ACME`

Another odd thing to make this work on pm-cpu:
With current master, we use a trick to distinguish between diff NERSC machines, but with this branch, I made the following hack, which would look for "login" as the login name.  This is working, but not ideal as it could conflict with a different machine also expecting "login".
```
  <!--NODENAME_REGEX>$ENV{NERSC_HOST}:perlmutter</NODENAME_REGEX-->
  <NODENAME_REGEX>login</NODENAME_REGEX>
```

I also found 2 issues with DEBUG attempts that can be avoided with 2 simple changes in e3sm source files.
Here are two places that were failing with runtime segfault in DEBUG due to the compiler looking-ahead
and trying to parse an experession B that should only be evaluated if A is true.
`components/mosart/src/riverroute/RtmMod.F90`
`components/cam/src/physics/cam/micro_mg2_0.F90`
We can decide how important fixes like these are. I don't see a fix in current master
at these locations, but it's possible the code flow doesn't happen same way in maint-1.0

Note: we may not have actually supported GNU compilers with maint-1.0.

I've added a `notes.txt` file to the branch temporarily to explain above and make it easier to make the changes.

**Comment by ndkeen** [comment]
For the build problem in `components/clm/src/external_models/sbetr/src/betr/betr_math/ODEMod.F90`, I pinged @peterdschwartz to see if we could make a change to maint-1.0 similar to https://github.com/E3SM-Project/E3SM/pull/4364

For the problem where we see `ERROR: picard convergence failed!` runtime error in MPAS ice, we need to do the same thing as https://github.com/E3SM-Project/E3SM/pull/4672
However, the CIME mechanisms to allow for conditional compiling of MPAS source files was not working in maint-1.0.
That is, I tried both of these:
```
#ice_shortwave.o: ice_shortwave.F90
#       $(FC) -c $(INCLDIR) $(INCS) $(FFLAGS) -fno-tree-pta $<
ice_shortwave.o: ice_shortwave.f90
        $(FC) -c $(INCLDIR) $(INCS) $(FFLAGS) -fno-tree-pta $<
```
and neither had an impact.
We could add this flag for all MPAS builds (as I show above in the notes.txt files) or add the flag to ALL fortran builds -- I doubt it will have performance impact, just seems unnecessary.

And, to better fix the issue of identifying the machine, we would need a CIME change noted here:
https://github.com/E3SM-Project/E3SM/pull/4773

**Comment by rljacob** [comment]
For the sbetr change, make a branch in sbetr called something like "e3sm-maint1.0", add the change to that branch, then update the submodule to point to that hash.

**Comment by ndkeen** [comment]
How about I create a PR to merge the branch, which gets us closer, then make another issue to sort out remaining issues?

**Comment by rljacob** [comment]
Yes go ahead.

**Comment by ndkeen** [comment]
Actually needed to update the module versions again. Will do some retesting and make a PR to get us much closer to having 1.0 working on pm-cpu after PM returns from sched maint.

**Comment by ndkeen** [comment]
I now have maint-1.0 working with Intel, which I can add to this branch.

There are 2 changes that needed to be made, and then some tests that still fail may have been failing before with Intel.

Some runtime errors can be resolved with a check for div-by-zero here `components/clm/src/biogeophys/SoilWaterMovementMod.F90`
I think it's safe to do this (copied from master).

The HOMME test tries to run but has an issue -- I see the openmp flag used is not correct
and I cannot find where/how it is applied.

And then several MPAS build fails can be resolved if I copy from master this file:
`components/mpas-source/src/framework/xml_stream_parser.c`

However, I don't know if that is OK to do -- and/or if other changes may be needed.
Plus that change is in mpas-source.

**Comment by rljacob** [comment]
Yes its ok to copy those.

**Comment by ndkeen** [comment]
How to make a change to mpas-source submodule?

**Comment by rljacob** [comment]
@jonbob we need to make modifcation to https://github.com/MPAS-Dev/MPAS-Model.git to get maint-1.0 working on pm-cpu.   How should we do that?  The hash its using is b273ceb622d74359.   Ideally, we would make a branch off of that commit called something like "e3sm-maint-1-0", add what we need, and point the submodule to that branch.  If we can't write to MPAS-Dev/MPAS-Model.git, then we'll have to fork it to E3SM-Project and make the mods there.

**Comment by jonbob** [comment]
@rljacob - let me look (and try to remember)

**Comment by jonbob** [comment]
We should be able to write there, and there is an e3sm/maint-1.0 branch to work from

**Comment by jonbob** [comment]
there should be documentation on this page: [Updating+MPAS+Framework+and+Components+in+ACME](https://acme-climate.atlassian.net/wiki/spaces/OCNICE/pages/95977592/Updating+MPAS+Framework+and+Components+in+ACME). I can help, but probably not until Monday if it doesn't make sense

**Comment by rljacob** [comment]
@jonbob Can you make sure Noel has write access?

**Comment by jonbob** [comment]
Noel is listed in the MPAS-Dev organization

**Comment by rljacob** [comment]
> How to edit/use a different branch of components/mpas-source in a maint-1.0 e3sm branch?
First add the changes you want to the MPAS-Dev/MPAS-Model repo on the branch Jon mentioned. and commit them.  That creates a new hash in that repo.  Then go to components/mpas-source and do "git reset --hard <hash>" using that new hash.  That changes what commit E3SM is pointing to.  To commit that change, cd up one dir and do "git add mpas-source; git commit".

**Comment by ndkeen** [comment]
I haven't tried to look for the PR that made the change to `components/mpas-source/src/framework/xml_stream_parser.c`, ie I'm just assuming it's OK and sufficient to copy it from master.

I'm here and can make the change?
```
login14% git branch
* e3sm/maint-1.0
  master
```

**Comment by ndkeen** [comment]
Thanks to @bishtgautam, one of the annoying errors with using GNU has been fixed in the sbetr submodule.

**Comment by ndkeen** [comment]
I think we can close as maint-1.0 now works with intel on pm-cpu. Many tests will work with GNU, but still need a few changes to get all tests working. One of them described here as a more contained issue:
https://github.com/E3SM-Project/E3SM/issues/5710
